designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,359 @@
+      6. [How Should We Handle an Unspecified Reason When Others are Specified?](#how-should-we-handle-an-unspecified-reason-when-others-are-specified)
+
+## User Scenarios 
+1. Users need the capability to schedule upgrades only during business hours or within more restricted time windows. Additionally, they require a system that doesn't compromise the cost savings from consolidation when upgrades are blocked due to drift.
+2. Users want to minimize workload disruptions during business hours but still want to be able to delete empty nodes throughout the day.  That is, empty can run all day, while limiting cost savings and upgrades due to drift to non-business hours only.
+
+See Less Made Up Scenarios here: 
+- https://github.com/kubernetes-sigs/karpenter/issues/924 
+- https://github.com/kubernetes-sigs/karpenter/issues/753#issuecomment-1790110838
+- https://github.com/kubernetes-sigs/karpenter/issues/672
+- https://github.com/kubernetes-sigs/karpenter/issues/1179 
+
+See further breakdown of some additional scenarios towards the bottom of the document 
+
+## Clarifying the requirements and behavior 
+**Reason and Budget Definition:** Users should be able to define an reason and a corresponding budget(s).
+**Supported Reasons:** All disruption Reasons affected by the current Budgets implementation (underutilized, empty, expired, drifted) should be supported. 
+**Default Behavior for Unspecified Reasons:** Budgets should continue to support a default behavior for all disruption reasons. 
+
+
+# API Design
+## Approach A: Extending the Budget API to specify a reason 
+This section contrasts two approaches for specifying disruption reasons in the v1beta1 nodepool API: a list of reasons (List Approach) and a single reason per budget (Single Reason Approach).
+
+### List Approach: Multiple Reasons per Budget - Recommended
+This approach allows specifying multiple disruption methods within a single budget entry. It is proposed to add a field Reasons to the budgets, which can include a list of reasons this budget applies to.
+#### Proposed Spec
+Add a simple field "reasons" is proposed to be added to the budgets. 
+```go
+// Budget defines when Karpenter will restrict the
+// number of Node Claims that can be terminating simultaneously.
+type Budget struct {
+      // Reasons is a list of disruption reasons. If Reasons is not set, this budget applies to all methods.
+      // If a reason is set, it will only apply to that method. If multiple reasons are specified,
+      // this budget will apply to all of them. If a reason is unspecified we will take the min value of this budget and the rest of the active budgets.
+      // if an unspecified reason exists we will also override all other reasons with its value if they are smaller than the unspecified reason.
+      // allowed reasons are "underutilized", "expired", "empty", "drifted"
+      // +kubebuilder:validation:MaxItems=5
```

Comment:
Why would it be 5 items? Is there a way to ensure values are unique?

Reply:
good catch we no longer have 5 items as default is gone. This should be 4.

Reply:
I haven't tested this, but the docs show we can use Unique items 


// +kubebuilder:validation:UniqueItems
bool
specifies that all items in this list must be unique.

Reply:
See: https://book.kubebuilder.io/reference/markers/crd-validation for reference